Hunger and starvation improvements

bolaft opened the pull request:
This PR addresses a few long-standing issues with hunger (see #11550, #11514):

- If you don't eat for a week, you do not need a week's worth of food to be satiated. 
- It's not possible to recover instantly from severe starvation just by having a very large meal.
- It should not be possible to eat a week's worth of food in one sitting.

In order to solve this, this PR separates the starvation logic from the hunger stat and links it to a newly introduced hidden stat, starvation. The process is simple: hunger is now capped at 400 (just above 'famished'). When hunger is at its maximum, starvation slowly increases. When hunger is NOT maxed out, starvation slowly decreases. Stat penalties and death are now tied to starvation, not just hunger. In terms of realism one could think of the starvation stat as an abstraction of body fat and how much nutrients are in the PC's system, while hunger is merely an indicator of the fullness of the stomach.

### Gameplay effects

These relatively small changes have a few interesting gameplay effects:

1) Starving characters can't eat thirty full meals in one sitting anymore. The amount of food a character can eat at once is limited to a much more reasonable amount. For example, previously it was technically possible to eat 500 bananas in a row. Now 34 bananas will be enough to fill up even the most banana-starved character.
2) Starvation effects remain for a while. Without this update, eating enough food can instantly bring a character from near death to perfect condition. With this update, a character who starved for a long time will not instantly lose all negative modifiers. Instead, the penalties will diminish and disappear gradually. Recovery time is roughly equal to the time spent starving (depending on metabolism modifiers).
3) Rationing food in times of scarcity now makes sense. For instance, previously a character who planned to hide for two weeks (for example, to get rid of a meth addiction) with only one week's worth of food had no incentive to ration it: whether all the food was eaten in the first week or spread across the two weeks made no difference in the end. With this update, it is advisable to spread out meals to slow down the starvation meter.

### Display

The player can see its character's starvation penalties in the character's page. They are displayed under the "Malnourished" effect. On the sidebar, the "hungry", "very hungry" and "famished" messages appear the same as before, depending on hunger. However, the "near starving" and "starving!" messages only appear if hunger is capped AND starvation is high enough. Therefore, a character with a full belly won't see those messages appear, even if its starvation is high. This reflects the fact that the character is not hungry, even though he or she still suffers from malnutrition penalties, and the player can't do anything about it at the moment anyway. As for the 'Hunger' speed modifier display, in order to avoid confusing players whose PC is not hungry (but still suffers from starvation), the label is changed to 'inanition'.

![screen_cdda_pr](https://user-images.githubusercontent.com/2310791/42344567-063d993e-809d-11e8-9b08-af4613c489e7.png)

### Balance

This PR does not affect established numbers. Characters will die from starvation exactly as fast as before (unless they ration food), and the speed, strength, dexterity etc. penalties from lack of food remain the same.

These changes should not affect balance much. They make the game a little bit easier by making it possible to ration food to alleviate starvation penalties, but they also make the game a little bit harder since it now takes some time to get rid of those penalties.

### Justification for the choice of 400 as a hunger cap

The only number I introduced in this PR is the 400 cap for nutrition. Here's why:

1) It's just above the current threshold for 'famished', which is the last threshold before the concept of starvation is introduced in the game's messages.
2) The world record for most bananas eaten in a row is 30. 30 bananas times 12 nutrition points per banana equals 360.
3) The game must offer players the opportunity to break the world's record for most bananas eaten in one sitting (pretty sure fans of the game would riot otherwise).
4) Current hunger thresholds are all rounded to the nearest hundred. 360 rounded to the nearest hundred is 400.

### Note

This is the first time I look into CDDA's code, and I don't usually use C++, so all comments and suggestions about the code are welcome. I tested the changes in game, they seem to work fine and everything behaves as expected, but I'd appreciate it if a CDDA veteran double checked to make sure I didn't break anything.

## Updates

See comments.

- Added "Severely Malnourished" effect.
- Added descriptive text to "Malnourished" and "Severely Malnourished".
- Lowered hunger cap from 400 to 300.
- Lowered "famished" threshold from 300 to 250 (otherwise it would never display).

Night-Pryanik commented:
Very interesting!

DracoGriffin commented:
Great addition, extremely clear reasoning and details. Very good work.

bolaft commented:
Thanks for the feedback!

Yeah I'm not too happy with "inanition" either actually. I didn't think of ESL people though, even though I am one myself. Initially I intended to use two differents modifiers, a "Hunger" speed modifier and a "Malnourished" modifier. However the `hunger_speed_penalty()` function is not linear, so hunger_speed_penalty(get_hunger()) + hunger_speed_penalty(get_starvation()) does not equal `hunger_speed_penalty(get_starvation() + get_hunger())`.

I could tweak it so there are two separate speed modifiers, but I think I would have a hard time getting exactly the same modifier values as before.

Or, I could just change "inanition" for something else, but I'm not sure what. "Lack of food" maybe?

draeath commented:
Could this be integrated with nutrition instead? They are already added/removed from food/lack-of-food, and decoupled from hunger itself. It would make sense that a "calorie" nutrient be added and used for this, instead of a new (hidden) statistic.

It would also make sense from a data perspective - food would have, in addition to nutrients, a caloric value. So long as this is kept reasonable to real values, this would make it intuitive for folks adding new foods.

I wonder if you might even scale the depletion based on player activity, as well.

bolaft commented:
@draeath Maybe I'm not understanding your suggestion right, but I don't really see how using the nutrition system for this would address the starvation issues? If "calories" became a nutrient similar to "Vitamin C" for example, it would not really solve the issue of the "bottomless pit stomach", nutrition needs are linear currently. So if a character misses a week's worth of calories, they would need to eat a week's worth of calories to go back to normal. Moreover, I don't think the nutrition system allows for gradual recovery: if a character has scurvy, taking three multivitamins fixes it instantly, and the same would go for calories: taking a week's worth of food at once would be required to fix it. And if the hunger is capped, then it would be impossible to keep up with caloric nutritional need, which would keep going down while the character is unable to eat any more.

Moreover, I think I remember reading on a comment thread that nutrition values are already based on calories, so adding a new "calorie nutrient" would be redundant. Furthermore, adding a new nutrient type that would need a value for every single foodstuff sounds like a lot of JSON editing work to me, and would mean that every mod adding food items would need to be updated, doesn't it?

Again I'm not sure I understood your suggestion correctly, please correct me if I'm wrong.

Vasyan2006 commented:
Nice addition, but I have some comments.
400 nutritions is too much for a hunger cap. Default character needs 288 nutritions per day. It is technically possible to eat 1.5 day rations in few minutes and not to vomit from overeating, but most of nutritions will be lost, because digestional tract cannot work at such speed. This value must be set by something like `stomach_capacity()`.
Overeating after long starving can be deadly. The weakened body cannot handle huge ammount of proteins, fats and sugar so this overloading will signifficantly damage methabolism even more. But not sure if such level of realism is required.
Bananas are certified to be used only as arbitrary dose of ionizing radiation (Banana equivalent dose). Using them as a metric of nutritions may lead to misunderstandings.

nexusmrsep commented:
Fantastic! Quick idea: Malnourished effect (and other effect levels if you plan them) should also have some descriptive text as a feedback for the player who might think "Satiated" for hunger and "Malnourished" for starvation as a potential bug.

Something along the lines of: "You haven't eaten for so long, that your body is weak, and you are a shadow of your former self. Only time and regular meals will help you recover."

Mecares commented:
I am not a expert, but is possible intake of calories really capped by amount of calories in reality or is it more a question of volume of food i can fit into my digestive system? Would making the choice between high energy density and low energy density food in a starving situation matter?

bolaft commented:
Ok, many good suggestions/comments in this thread. So to sum up:

- The "Inanition" label should be changed for something else (@DracoGriffin)
- Integrating this with the nutrition system instead (@dreath)
- Lowering the 400 threshold (@Vasyan2006)
- Overeating after a long starvation period can be deadly (@Vasyan2006)
- Adding descriptive text for the "Malnourished" effect (@nexusmrsep)
- 'Hunger' as in, the *feeling* of hunger, and how much food is in the stomach are different things (@kevingranade)
- Calorie intake might be based on nutritional value, while max stomach content might be based on volume (@Mecares)

So for this PR I don't think I should go too much into the complexities of realistic starvation.

I added a commit to the PR based on @DracoGriffin's suggestions:

- The "Malnourished" effect now has descriptive text ("Your body is weakened by starvation. Only time and regular meals will help you recover."). I didn't use the "You haven't eaten for so long" part of the suggestion because it is possible to be malnourished even though the character just ate a meal.
- I added a "Severely Malnourished" effect. It works the same, but shows up when starvation is very advanced. The text message reflects this ("Your body is severely weakened by starvation. You might die if you don't start eating regular meals!").

A couple questions:

- Do I change the "Inanition" speed modifier label? To what?
- Do I lower the Hunger cap? To what?
- Should I make special rules for special situations? At the moment, I made it so taking mutagen affects starvation directly if hunger is high enough, but there are other cases in which such effects could be debatable (such as plant mutants etc.)

Some of the other suggestions might make for an interesting separate PR:

- Separating "hunger" from "stomach contents". Hunger would only be a subjective indicator, subject to special effects (like marijuana), only used for display. Lack of actual food in the stomach would be the actual trigger for starvation and penalties.
- Stomach content would be limited by food volume, rather than nutritional value, making it smart to eat high density foods before leaving a long trip.
- Maybe other realism improvements, but I'm not sure it would be the best idea. It would add too much complexity for too little gameplay benefits IMO.

DracoGriffin commented:
Clarification is probably warranted in here regarding the difference between food nutritional value (ie., vitamins) and food nutrition which isn't the best name for "calories".

>I am not a expert, but is possible intake of calories really capped by amount of calories in reality or is it more a question of volume of food i can fit into my digestive system? Would making the choice between high energy density and low energy density food in a starving situation matter?

Volume of food is a super slippery slope and [extremely variable](https://hypertextbook.com/facts/2000/JonathanCheng.shtml) in human beings already, notwithstanding mutations and CBMs and everything else.

In terms of starvation recovery (like, actual starvation), is that [you can harm yourself from overindulging](https://en.wikipedia.org/wiki/Refeeding_syndrome) (refeeding syndrome), even though "overindulgence" amount may look like a normal size meal for other people.

So for example, my first link shows the average human stomach can stretch to 4 liters, yet how much could a starved person's stomach contain? Also, then you gotta consider [extreme cases like this one](https://en.wikipedia.org/wiki/Nathan%27s_Hot_Dog_Eating_Contest#By_year). So how will this all affect survivors with traits/mutations that affect metabolism and such -- do bovine post-thresholds still only eat this volume or something else?

You bring up a good point. The nutritional and caloric density value of a food is also important -- some foods would be much more suitable for starvation recovery and such.

Anywho, there's my shotgun blast of information. And lastly, an [interesting food calculator](https://www.aqua-calc.com/calculate/food-volume-to-weight) that spits out more information you could ever want. (which just calls from [USDA Food Composition Databases](https://ndb.nal.usda.gov/ndb/search/list) but converts some of the information into a more ... *digestible* format.)

For funsies, here's what one liter (1 L) of sliced bananas offers:

From | kilocalories(kcal) | kilojoule(kJ)
-- | -- | --
Carbohydrate | 521.31 | 2 181.16
Fat | 17.51 | 73.27
Protein | 23.22 | 97.15
Other | 2.23 | 9.32
Total | 564.27 | 2 360.91


NutrientFind foods rich in specific nutrients | Unit | Value per634.01 g
-- | -- | --
Proximates
Water | g | 474.94
Energy | kcal | 564
Protein | g | 6.91
Total lipid (fat) | g | 2.09
Carbohydrate, by difference | g | 144.81
Fiber, total dietary | g | 16.5
Sugars, total | g | 77.54
Minerals
Calcium, Ca | mg | 32
Iron, Fe | mg | 1.65
Magnesium, Mg | mg | 171
Phosphorus, P | mg | 139
Potassium, K | mg | 2270
Sodium, Na | mg | 6
Zinc, Zn | mg | 0.95
Vitamins
Vitamin C, total ascorbic acid | mg | 55.2
Thiamin | mg | 0.197
Riboflavin | mg | 0.463
Niacin | mg | 4.216
Vitamin B-6 | mg | 2.327
Vitamin B-12 | μg | 0
Vitamin A, RAE | μg | 19
Vitamin A, IU | IU | 406
Lipids
Fatty acids, total saturated | g | 0.71
Fatty acids, total monounsaturated | g | 0.203
Fatty acids, total polyunsaturated | g | 0.463
Cholesterol | mg | 0
Other
Caffeine | mg | 0

DracoGriffin commented:
>The "Inanition" label should be changed for something else (@DracoGriffin)

Honestly, keep it because it is extremely accurate (`exhaustion from lack of nourishment; starvation`). I was just saying out of all the work you did, that was really my only "non-issue" with it. But there really isn't a better descriptor that doesn't obfuscate other similar terms already being used (just like Malnourished can be confused with lack of food, lack of eating, or lack of vitamins, but we really should make these terms more stringent -- starvation and adjectives, hungry and adjectives, and hypovitaminosis and adjectives for more specific vitamin deficiencies, respectively -- malnourished could fall under starvation or hypovitaminosis, but then it'd be more appropos to use the term malnutrition).

bolaft commented:
> just like Malnourished can be confused with lack of food, lack of eating, or lack of vitamins, but we really should make these terms more stringent -- starvation and adjectives, hungry and adjective (...)

Yes maybe it would be better to do away with it, but the reason why I used "malnourished" rather than "starving" or "starvation" is because I felt the latter implies the character is _currently_ starving, rather than possibly suffering from the effects of a previous period of starvation.

But I can change it.

DracoGriffin commented:
>Yes maybe it would be better to do away with it, but the reason why I used "malnourished" rather than "starving" or "starvation" is because I felt the latter implies the character is currently starving, rather than possibly suffering from the effects of a previous period of starvation.

Keep those too, Malnourished/Severely Malnourished also make sense in this context. We already have [hypervitaminosis and the appropriate hypovits here](https://github.com/CleverRaven/Cataclysm-DDA/blob/e85269de44d5c5bcc084f89fd343d9573f915bef/data/json/vitamin.json#L14).

bolaft commented:
I removed this comment previously but since you answered it I put it back so the conversation makes sense:

> just like Malnourished can be confused with lack of food, lack of eating, or lack of vitamins, but we really should make these terms more stringent -- starvation and adjectives, hungry and adjective (...)

Yes maybe it would be better to do away with it, but the reason why I used "malnourished" rather than "starving" or "starvation" is because I felt the latter implies the character is _currently_ starving, rather than possibly suffering from the effects of a previous period of starvation.

But I can change it.

bolaft commented:
`gorgon-ghprb` and `continuous-integration/appveyor/pr` fail, what do those check for? What are "astyle regressions"? gorgon-ghprb finds weird errors, like: 

```
..\src\sidebar.cpp(318): error C2146: syntax error: missing ';' before identifier 'get_starvation' [C:\Projects\Cataclysm-DDA\msvc-full-features\Cataclysm.vcxproj]
```

But the line is : 
```
    if( get_hunger() >= 400 and get_starvation() > 2400 ) {
```
Putting a `;` before `get_starvation` would make no sense.

(I have no trouble compiling on my system).

DracoGriffin commented:
>1. Do I change the "Inanition" speed modifier label? To what?
>2. Do I lower the Hunger cap? To what?
>3. Should I make special rules for special situations? At the moment, I made it so taking mutagen affects starvation directly if hunger is high enough, but there are other cases in which such effects could be debatable (such as plant mutants etc.)

1. So, keep "Inanition". That's done.

2. I'd suggest changing the hunger cap to 300 due to values being chosen on [this line of information/code](https://github.com/CleverRaven/Cataclysm-DDA/blob/bb41bf6c6e10344617941be565b1c02e76ca783a/src/itype.h#L166):
> /** 1 nutr ~= 8.7kcal (1 nutr/5min = 288 nutr/day at 2500kcal/day) */

so that 12 nutrition can fluctuate for the player in terms of under/overeating if they choose.

And lastly, there are quite a few variables so I don't know them all but one that might be an issue is [metabolism_modifier](https://github.com/CleverRaven/Cataclysm-DDA/blob/a593644c490aceeb21da58361d73ad4ecc6d8132/src/consumption.cpp#L262).

3. I'd suggest making adjustments that the hunger cap is increased half of the metabolism_modifier value so characters with those traits don't end up rapidly starving to death. I am probably misunderstanding something, though.

ZhilkinSerg commented on src/sidebar.cpp:
```diff
@@ -315,9 +315,9 @@ void player::disp_status( const catacurses::window &w, const catacurses::window
 
     wmove( w, sideStyle ? 1 : 2, 0 );
-    if( get_hunger() > 2800 ) {
+    if( get_hunger() >= 400 and get_starvation() > 2400 ) {
```
Use `&&` instead of `and` here as logical AND.

ZhilkinSerg commented on src/sidebar.cpp:
```diff
@@ -315,9 +315,9 @@ void player::disp_status( const catacurses::window &w, const catacurses::window
+    if( get_hunger() >= 400 and get_starvation() > 2400 ) {
         wprintz( w, c_red,    _( "Starving!" ) );
-    } else if( get_hunger() > 1400 ) {
+    } else if( get_hunger() >= 400 and get_starvation() > 1000 ) {
```
Use `&&` instead of `and` here as logical AND.

bolaft replied:
Will do.

ZhilkinSerg replied:
This should be one of the reasons why VS build is failing in AppVeyor - VC++ doesn't respect standards.

bolaft commented:
New commits :

- Lowered hunger cap from 400 to 300.
- Lowered "famished" threshold from 300 to 250 (otherwise it would never display).
- Code style fixes ("and" => "&&")
- Corrected how mutations affect starvation: starvation is directly updated when mutagen is consumed, not when mutations are activated.

bolaft commented:
> I'd suggest making adjustments that the hunger cap is increased half of the metabolism_modifier value so characters with those traits don't end up rapidly starving to death.

Hum. I'm conflicted. On one hand, it makes sense that people with a fast metabolism would start starving faster, and would need to eat more regularly. I don't see a problem with that, actually I like it. However, it's true that some modifications to metabolism come from animal mutation trees, and it would make sense that a half bear mutant should be able to eat a lot at once.

Maybe the problem could be solved by modifying the "get_stomach_capacity" function. Currently it allows survivors with the "Hibernation" and "Gourmand" traits to eat more than their fill. I could add "Fast Metabolism" and "Very Fast Metabolism" to the list.

bolaft commented:
Another thing: for now in this PR, I made it so starvation increases and decreases over time at the same rate as hunger. It's simple and it works, but it might not be the most logical solution in some cases. For example a character with a fast metabolism recovers from starvation faster, a sleeping character recovers from starvation slower. If you feel it should work differently I'm taking suggestions.

DracoGriffin commented:
>Hum. I'm conflicted. On one hand, it makes sense that people with a fast metabolism would start starving faster, and would need to eat more regularly. I don't see a problem with that, actually I like it. 

I agree. This is probably relating to my misunderstanding of how the mechanic works exactly -- that the hunger cap was based upon a day/reset, so you couldn't eat more than a day's worth of calories.

>However, it's true that some modifications to metabolism come from animal mutation trees, and it would make sense that a half bear mutant should be able to eat a lot at once. Maybe the problem could be solved by modifying the "get_stomach_capacity" function. Currently it allows survivors with the "Hibernation" and "Gourmand" traits to eat more than their fill. I could add "Fast Metabolism" and "Very Fast Metabolism" to the list.

It would potentially be better to hook them into the post-threshold (since shedding humanity basically) or something, rather than the metabolism itself. A higher basal metabolic rate doesn't necessarily mean a larger capacity for food.

bolaft commented:
> It would potentially be better to hook them into the post-threshold (since shedding humanity basically) or something, rather than the metabolism itself.

Yes, good idea. But it would be strange to give a bonus to a survivor who does not have the metabolic mutation, even though he/she is post-threshold, so I'm thinking about making a bonus to stomach capacity based on whether the PC has both a metabolic mutation and is post-threshold. I'm thinking +50 for "Rapid Metabolism", +75 for "Fast Metabolism", and +100 for "Very Fast Metabolism", and apply it to post-threshold characters of the Beast, Slime, Raptor, Chimera and Rat mutation trees.

That would also have the effect of displaying "sated" and "gorged" hunger status for these character. Currently only hibernating bear mutants can see them, I'm not sure if there's a deeper reason for it.

Does that sound good?

Mecares commented:
I personally even think the stats a player has should affect hunger generation, the mountains of muscles some player characters have need some energy to be maintained after all.

bolaft commented:
Well if we limit this PR to basic functionality (which I think is a good idea), then it's pretty much done, I don't see what else needs to change.

I tested it for a several hours on my current playthrough without issues. I also did several tests on the debug mod. The only odd thing I noticed was that gradual changes were slightly off (I tested it by giving a survivor a lot of fatigue and making him sleep for over three days):

Without PR:
- Hunger 475
- Thirst 476

With PR:
- Hunger 300 + Starvation 180 = 480 (=> 1% difference)
- Thirst 497 (=> 4% difference)

I looked into it and I'm not sure what is causing these differences, especially concerning thirst since I didn't touched thirst at all. Debug messages show the same metabolic rate and the same thirst rate. I'll try to figure it out by displaying more variables as soon if I have time today.

nexusmrsep commented on src/iuse.cpp:
```diff
@@ -1161,9 +1161,16 @@ int iuse::mutagen( player *p, item *it, bool, const tripoint & )
             p->mod_thirst( m_category.mutagen_thirst );
             p->mod_fatigue( m_category.mutagen_fatigue );
+
+            if ( m_category.mutagen_hunger + p->get_hunger() > 300 ) {
```
astyle doesn't want white spaces between 'if' and '('

nexusmrsep commented:
Astyle regression means code style is not in standard for this repository. See `CODE_STYLE.md` in docs.
It's sometimes hard to see style "errors" by a visual inspection, so AStyle comes to the rescue.
Jenkins' log shows what files are affected.

Download your branch locally so you have it on you harddrive. From a command line run AStyle binnary from `tools` folder using `CODE_STYLE.md` parameters and target files: src/iuse.cpp, src/player_display.cpp, src/consumption.cpp. Inspect changes and commit them. I personaly use GitHub Desktop to control my local branches, as it has intuitive visual interface.

bolaft commented:
Ok I will do that soon, thanks for the explanation. I'm a bit busy this weekend but worst case scenario I'll push a commit monday.

cainiaowu commented:
Starvation texts need to be translateable.

cainiaowu commented:
Yeah, currently on a patato phone so I cant comment on specific line right now.

bolaft commented:
New commit: made starvation descriptions translatable.

bolaft commented:
I did more testing, the small discrepancies I found earlier are not an issue, there is an element of randomness to hunger/thirst increase. Two tests on the version of the game without this PR also give slightly different numbers (~5% difference). So as far as I can tell everything behaves as expected, I didn't find any bugs.

I think this PR is ready.

Too-DAMN-Much commented:
so either how do i disable this in my game or when are we getting fat for calorie batteries?

Night-Pryanik commented:
> how do i disable this in my game

You make your own fork, you get compiler, you disable all that you don't like, you compile the game.

Too-DAMN-Much commented:
@Night-Pryanik oh well, no loss regardless, it's been over 4 months since i even bothered to update the game it's been so unplayable and crashy anyway, i literally can't reliably play for 20 minutes without crashing last time i tried so cata is on permanent have given up status for me anyway.
